src/main/java/backend/Logic.java

```diff
@@ -200,18 +198,77 @@ public Model getRepo(String repoId) {
      * @return true if label replacement on GitHub was a success, false otherwise.
      */
     public CompletableFuture<Boolean> replaceIssueLabels(TurboIssue issue, List<String> newLabels) {
+        List<String> originalLabels = issue.getLabels();
+
         logger.info("Changing labels for " + issue + " on UI");
-        issue.setLabels(newLabels);
+        /* Calls models to replace the issue's labels locally since the the reference to the issue here
+           could be invalidated by changes to the models elsewhere */
+        Optional<TurboIssue> localReplaceResult =
+                models.replaceIssueLabels(issue.getRepoId(), issue.getId(), newLabels);
+        if (!localReplaceResult.isPresent()) {
+            return CompletableFuture.completedFuture(false);
+        }
         refreshUI();
 
-        return updateIssueLabelsOnRepo(issue, newLabels);
+        return updateIssueLabelsOnServer(issue, newLabels)
+                .thenApply((isUpdateSuccessful) -> handleIssueLabelsUpdateOnServerResult(
+                            isUpdateSuccessful, localReplaceResult.get(), originalLabels));
+    }
+
+    /**
+     * Gets the issue identified by {@code repoId} and {@code issueId} in {@link Logic#models}
+     * @param repoId
+     * @param issueId
+     * @return
+     */
+    private Optional<TurboIssue> getIssue(String repoId, int issueId) {
```

Comment:
Should this be a MultiModel method instead? Seeing as it only operates on `models`.

Also, would a caller ever expect to handle the empty case (can the issue be missing in some legitimate case)? If not, maybe it should throw a runtime exception upon failure, since that probably indicates something seriously wrong. I think this is a problem with other MultiModel methods as well.